go/cli/mcap/cmd/sort.go

```diff
@@ -0,0 +1,196 @@
+		return errUnindexedFile{err}
+	}
+
+	isEmpty, err := fileHasNoMessages(r)
+	if err != nil {
+		return fmt.Errorf("failed to check if file is empty: %w", err)
+	}
+
+	if len(info.ChunkIndexes) == 0 && !isEmpty {
+		return errUnindexedFile{errors.New("no chunk index records")}
+	}
+
+	err = writer.WriteHeader(info.Header)
+	if err != nil {
+		return fmt.Errorf("failed to write header: %w", err)
+	}
+
+	// handle the attachments and metadata metadata first; physical location in
+	// the file is irrelevant but order is preserved.
+	for _, index := range info.AttachmentIndexes {
+		attReader, err := reader.GetAttachmentReader(index.Offset)
+		if err != nil {
+			return fmt.Errorf("failed to read attachment: %w", err)
+		}
+		err = writer.WriteAttachment(&mcap.Attachment{
+			Name:       index.Name,
+			MediaType:  index.MediaType,
+			CreateTime: index.CreateTime,
+			LogTime:    index.LogTime,
+			DataSize:   index.DataSize,
+			Data:       attReader.Data(),
+		})
+		if err != nil {
+			return fmt.Errorf("failed to write attachment: %w", err)
+		}
+	}
+	for _, index := range info.MetadataIndexes {
+		metadata, err := reader.GetMetadata(index.Offset)
+		if err != nil {
+			return fmt.Errorf("failed to read metadata: %w", err)
+		}
+		err = writer.WriteMetadata(metadata)
+		if err != nil {
+			return fmt.Errorf("failed to write metadata: %w", err)
+		}
+	}
+
+	it, err := reader.Messages(mcap.UsingIndex(true), mcap.InOrder(mcap.LogTimeOrder))
+	if err != nil {
+		return fmt.Errorf("failed to read messages: %w", err)
+	}
+	schemas := make(map[uint16]*mcap.Schema)
+	channels := make(map[uint16]*mcap.Schema)
+	for {
+		schema, channel, message, err := it.Next(nil)
+		if err != nil {
+			if errors.Is(err, io.EOF) {
+				break
+			}
+		}
+		if schema != nil {
+			if _, ok := schemas[schema.ID]; !ok {
+				err := writer.WriteSchema(schema)
+				if err != nil {
+					return fmt.Errorf("failed to write schema: %w", err)
+				}
+			}
+		}
+		if _, ok := channels[channel.ID]; !ok {
+			err := writer.WriteChannel(channel)
+			if err != nil {
+				return fmt.Errorf("failed to write channel: %w", err)
+			}
+		}
+		err = writer.WriteMessage(message)
+		if err != nil {
+			return fmt.Errorf("failed to write message: %w", err)
+		}
+	}
+
+	return writer.Close()
+}
+
+var sortCmd = &cobra.Command{
+	Use:   "sort [file] -o output.mcap",
+	Short: "Read an MCAP file and write the messages out physically sorted on log time",
+	Run: func(cmd *cobra.Command, args []string) {
+		if len(args) != 1 {
+			die("supply a file")
+		}
+		f, err := os.Open(args[0])
+		if err != nil {
+			die("failed to open file: %s", err)
+		}
+		defer f.Close()
+
+		output, err := os.Create(sortOutputFile)
+		if err != nil {
+			die("failed to open output: %s", err)
+		}
+		err = sortFile(output, f)
+		if err != nil {
+			if errors.Is(err, errUnindexedFile{}) {
+				die("Error reading file index: %s. "+
+					"You may need to run `mcap recover` if the file is corrupt or unchunked.", err)
```

Comment:
nit: should this be "unindexed" rather than unchunked? That would match the `errUnindexedFile` we're testing for

Reply:
unindexed doesn't seem right to me either, because we can have other kinds of indexes just not chunks. I can say chunk-indexed.